src/std/make.ss

```diff
@@ -20,224 +34,387 @@
+  (def nspec (vector-length spec@)) ; total number of targets
+  (def (specvec) (make-vector nspec #f)) ; vector indexed by target number
+  (def spec-index% (invert-hash<-vector spec@)) ; target number by spec
+  (def id@ (specvec)) ; dependency id by target number
+  (def id-index% (hash)) ; target number by dependency id
+  (def deps@ (specvec)) ; list of dependencies (as symbols) by target number
+  (def build-list []) ; actual build order by target number
+  (def timestamp% (hash)) ; timestamp by dependency id
+  (def blocked-by@ (specvec)) ;; #f or hash of numbers of specs blocked by the numbered target
+  (def blocking@ (specvec)) ;; #f or hash of numbers of specs blocking the numbered target
+  (def ready (make-pqueue identity <)) ;; targets that are ready to build, prioritized by order in user
+  (when (verbose>=? 3) (writeln [Step: 1 spec@: spec@]))
+
+  ;;; 2. Reinstate those cached dependencies from previous build that are still up-to-date:
+
+  (def (dependency-timestamp dep) ;; Get timestamp for each dependency by id symbol.
+    ;; We must be careful to only call that for modules outside the current project
+    (with-catch (lambda (_) +inf.0)
+                (lambda () (hash-ensure-ref timestamp% dep (cut library-timestamp dep)))))
+
+  (alet ((build-deps-timestamp (modification-time/cache build-deps-file))
+         (previous@ (read-build-deps build-deps-file)))
+    (def previous-out-of-date% (hash))
+    (def (previous-file-up-to-date? file)
+      (<= (file-timestamp file) build-deps-timestamp))
+    (def (previous-dependency-up-to-date? dep)
+      (or (hash-get spec-index% dep) ;; NB: we depend on the previous build-deps being topologically sorted
+          (and (not (hash-get previous-out-of-date% dep))
+               (<= (dependency-timestamp dep) build-deps-timestamp))))
+    (def (previous-entry-up-to-date? spec deps)
+      (and (andmap previous-file-up-to-date? (spec-inputs spec settings))
+           (andmap previous-dependency-up-to-date? deps)))
+    (def (consider-previous id spec deps)
+      (when (verbose>=? 7) (writeln [previous: id spec: spec deps: deps target: (hash-get spec-index% spec) up-to-date?: (and (hash-get spec-index% spec) (previous-entry-up-to-date? spec deps))]))
+      (def target (hash-get spec-index% spec))
+      (if (and target (previous-entry-up-to-date? spec deps))
+        (begin (vector-set! id@ target id)
+               (hash-put! id-index% id target)
+               (vector-set! deps@ target deps))
+        (hash-put! previous-out-of-date% id #t)))
+    (for-each (cut apply consider-previous <>) previous@))
+  (when (verbose>=? 3)
+    (writeln [Step: 2 previous-up-to-date: (hash->list/sort id-index% symbol<?)]))
+
+  ;; 3. Compute dependencies for entries that are out of date, or at least not cached
+  ;; NB: We assume this will catch any circular dependency and error out.
+  ((cut vector-for-each <> spec@)
```

Comment:
this is very funky.

Reply:
Not sure what you mean by "funky". I'm iterating over `spec@`, and making all the simple information obvious in a line, instead of requiring the reader to go 9 lines deep and back to understand what it's all about.

Reply:
Instead of writing `(vector-for-each proc spec@)` you write `((cut vector-for-each <> spec@) proc)`.
This is funky!
But it's ok, I can live with it.